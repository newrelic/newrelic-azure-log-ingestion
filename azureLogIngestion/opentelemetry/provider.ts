import { Resource } from "@opentelemetry/resources"
import { BasicTracerProvider } from "@opentelemetry/tracing"

export class NRTracerProvider extends BasicTracerProvider {
    // @ts-ignore
    resource: Resource

    updateResource(resource: Resource): void {
        this.resource = resource
    }
}
