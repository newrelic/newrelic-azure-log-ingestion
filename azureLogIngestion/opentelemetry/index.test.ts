jest.useFakeTimers()

import OpenTelemetryAdapter from "./index"

import {
    appInsightsAppDependency,
    appInsightsAppEvent,
    appInsightsAppException,
    appInsightsAppRequest,
    arrayOfStrings,
} from "../adapter/testdata.test"
import { BatchSpanProcessor } from "@opentelemetry/sdk-trace-base"

import { NRTracerProvider } from "./provider"

process.env["OTEL"] = "true"
process.env["otelJestTests"] = "true"

describe("OpenTelemetryAdapter", () => {
    let log
    let mockContext
    beforeAll(() => {
        log = (...args) => null
        log.verbose = (...args) => null
        log.info = (...args) => null
        log.warn = (...args) => null
        log.error = (...args) => null

        mockContext = {
            bindings: {},
            bindingData: {},
            bindingDefinitions: [],
            done: () => null,
            executionContext: { invocationId: "foobar", functionName: "foobar", functionDirectory: "foobar" },
            invocationId: "foobar",
            log,
            traceContext: { attributes: {}, traceparent: "foobar", tracestate: "foobar" },
        }
    })

    it("instantiates with api key", () => {
        const adapter = new OpenTelemetryAdapter("mock-insert-key", mockContext)

        expect(adapter).toBeInstanceOf(OpenTelemetryAdapter)

        expect(adapter.spanProcessor.spanProcessor).toBeInstanceOf(BatchSpanProcessor)
        expect(adapter.spanProcessor.batch).toBeInstanceOf(Array)
        expect(adapter.spanProcessor.batch.length).toEqual(0)
    })

    it("processes app request and dependency as spans", () => {
        const adapter = new OpenTelemetryAdapter("mock-insert-key", mockContext)

        expect(adapter.spanProcessor.batch.length).toEqual(0)
        expect(adapter.spanProcessor.batch).toMatchSnapshot()

        adapter.processMessages(appInsightsAppRequest, mockContext)
        expect(adapter.spanProcessor.batch.length).toEqual(2)
        expect(adapter.spanProcessor.batch).toMatchSnapshot()

        adapter.processMessages(appInsightsAppDependency, mockContext)
        expect(adapter.spanProcessor.batch.length).toEqual(4)
        expect(adapter.spanProcessor.batch).toMatchSnapshot()
    })

    it("processes app request and dependency when sent as array of strings", () => {
        const adapter = new OpenTelemetryAdapter("mock-insert-key", mockContext)

        expect(adapter.spanProcessor.batch.length).toEqual(0)
        expect(adapter.spanProcessor.batch).toMatchSnapshot()

        adapter.processMessages(arrayOfStrings, mockContext)
        expect(adapter.spanProcessor.batch.length).toEqual(6)
        expect(adapter.spanProcessor.batch).toMatchSnapshot()

        adapter.processMessages(appInsightsAppDependency, mockContext)
        expect(adapter.spanProcessor.batch.length).toEqual(8)
        expect(adapter.spanProcessor.batch).toMatchSnapshot()
    })

    it("processes app event as span", () => {
        const adapter = new OpenTelemetryAdapter("mock-insert-key", mockContext)

        expect(adapter.spanProcessor.batch.length).toEqual(0)
        expect(adapter.spanProcessor.batch).toMatchSnapshot()

        adapter.processMessages(appInsightsAppEvent, mockContext)

        expect(adapter.spanProcessor.batch.length).toEqual(2)
        expect(adapter.spanProcessor.batch).toMatchSnapshot()
    })

    it("processes app exception as span", () => {
        const adapter = new OpenTelemetryAdapter("mock-insert-key", mockContext)

        expect(adapter.spanProcessor.batch.length).toEqual(0)
        expect(adapter.spanProcessor.batch).toMatchSnapshot()

        adapter.processMessages(appInsightsAppException, mockContext)

        expect(adapter.spanProcessor.batch.length).toEqual(2)
        expect(adapter.spanProcessor.batch).toMatchSnapshot()
    })
})
