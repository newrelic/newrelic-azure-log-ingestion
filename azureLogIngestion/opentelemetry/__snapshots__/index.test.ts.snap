// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`OpenTelemetryAdapter processes app event as span 1`] = `Array []`;

exports[`OpenTelemetryAdapter processes app event as span 2`] = `
Array [
  Object {
    "attributes": Object {
      "name": "my custom event",
      "parentSpanId": "12345",
    },
    "duration": undefined,
    "endTime": Array [
      NaN,
      NaN,
    ],
    "ended": undefined,
    "events": Array [],
    "instrumentationLibrary": Object {
      "name": "default",
      "version": undefined,
    },
    "kind": 0,
    "links": Array [],
    "name": "my custom event",
    "parentSpanId": undefined,
    "resource": Resource {
      "attributes": Object {
        "service.name": "unknown_service:/Users/mrickard/.nvm/versions/node/v12.19.0/bin/node",
        "telemetry.sdk.language": "nodejs",
        "telemetry.sdk.name": "opentelemetry",
        "telemetry.sdk.version": "0.23.0",
      },
    },
    "spanContext": undefined,
    "startTime": Array [
      1620155556185,
      0,
    ],
    "status": Object {
      "code": 1,
    },
  },
]
`;

exports[`OpenTelemetryAdapter processes app exception as span 1`] = `Array []`;

exports[`OpenTelemetryAdapter processes app exception as span 2`] = `
Array [
  Object {
    "attributes": Object {
      "name": "at module.exports (C:/home/site/wwwroot/functionName/index.js:266:43)",
      "parentSpanId": "{long operation id}",
    },
    "duration": undefined,
    "endTime": Array [
      NaN,
      NaN,
    ],
    "ended": undefined,
    "events": Array [
      Object {
        "attributes": Object {
          "exception.message": "Error: Throwing an uncaught error on purpose.",
          "exception.stacktrace": Array [
            Object {
              "error.message": "Error: Throwing an uncaught error on purpose.",
              "id": "0",
              "outerId": "0",
              "parsedStack": Array [
                Object {
                  "assembly": "at module.exports (path/to/file:lineNumber:charPos)",
                  "fileName": "path/to/file.js",
                  "level": 0,
                  "line": 266,
                  "method": "module.exports",
                },
                Object {
                  "assembly": "at WorkerChannel.invocationRequest (path/to/file:lineNumber:charPos)",
                  "fileName": "path/to/file.js",
                  "level": 1,
                  "line": 18579,
                  "method": "WorkerChannel.invocationRequest (path/to/file:lineNumber:charPos)",
                },
                Object {
                  "assembly": "at ClientDuplexStream.methodName (path/to/file:lineNumber:charPos)",
                  "fileName": "path/to/file:lineNumber:charPos",
                  "level": 2,
                  "line": 18359,
                  "method": "ClientDuplexStream.methodName (Drive:/Directory",
                },
                Object {
                  "assembly": "at ClientDuplexStream.emit (path/to/file:lineNumber:charPos)",
                  "fileName": "file.js",
                  "level": 3,
                  "line": 314,
                  "method": "ClientDuplexStream.emit",
                },
              ],
              "severityLevel": "Error",
              "type": "Error",
            },
          ],
          "exception.type": "at module.exports (C:/home/site/wwwroot/functionName/index.js:266:43)",
        },
        "name": "exception",
        "time": Array [
          1620233096802,
          0,
        ],
      },
    ],
    "instrumentationLibrary": Object {
      "name": "default",
      "version": undefined,
    },
    "kind": 0,
    "links": Array [],
    "name": "at module.exports (C:/home/site/wwwroot/functionName/index.js:266:43)",
    "parentSpanId": undefined,
    "resource": Resource {
      "attributes": Object {
        "service.name": "unknown_service:/Users/mrickard/.nvm/versions/node/v12.19.0/bin/node",
        "telemetry.sdk.language": "nodejs",
        "telemetry.sdk.name": "opentelemetry",
        "telemetry.sdk.version": "0.23.0",
      },
    },
    "spanContext": undefined,
    "startTime": Array [
      1620233096802,
      0,
    ],
    "status": Object {
      "code": 2,
      "message": "Error: Throwing an uncaught error on purpose.",
    },
  },
]
`;

exports[`OpenTelemetryAdapter processes app request and dependency as spans 1`] = `Array []`;

exports[`OpenTelemetryAdapter processes app request and dependency as spans 2`] = `
Array [
  Object {
    "attributes": Object {
      "name": "{function name}",
      "parentSpanId": "12345",
    },
    "duration": undefined,
    "endTime": Array [
      1619736074132,
      159179688,
    ],
    "ended": undefined,
    "events": Array [],
    "instrumentationLibrary": Object {
      "name": "default",
      "version": undefined,
    },
    "kind": 0,
    "links": Array [],
    "name": "{function name}",
    "parentSpanId": undefined,
    "resource": Resource {
      "attributes": Object {
        "service.name": "unknown_service:/Users/mrickard/.nvm/versions/node/v12.19.0/bin/node",
        "telemetry.sdk.language": "nodejs",
        "telemetry.sdk.name": "opentelemetry",
        "telemetry.sdk.version": "0.23.0",
      },
    },
    "spanContext": undefined,
    "startTime": Array [
      1619736074123,
      0,
    ],
    "status": Object {
      "code": 1,
    },
  },
]
`;

exports[`OpenTelemetryAdapter processes app request and dependency as spans 3`] = `
Array [
  Object {
    "attributes": Object {
      "name": "{function name}",
      "parentSpanId": "12345",
    },
    "duration": undefined,
    "endTime": Array [
      1619736074132,
      159179688,
    ],
    "ended": undefined,
    "events": Array [],
    "instrumentationLibrary": Object {
      "name": "default",
      "version": undefined,
    },
    "kind": 0,
    "links": Array [],
    "name": "{function name}",
    "parentSpanId": undefined,
    "resource": Resource {
      "attributes": Object {
        "service.name": "unknown_service:/Users/mrickard/.nvm/versions/node/v12.19.0/bin/node",
        "telemetry.sdk.language": "nodejs",
        "telemetry.sdk.name": "opentelemetry",
        "telemetry.sdk.version": "0.23.0",
      },
    },
    "spanContext": undefined,
    "startTime": Array [
      1619736074123,
      0,
    ],
    "status": Object {
      "code": 1,
    },
  },
  Object {
    "attributes": Object {
      "name": "External/{name property of client.trackDependency}",
      "parentSpanId": "12345",
    },
    "duration": undefined,
    "endTime": Array [
      1620165671171,
      0,
    ],
    "ended": undefined,
    "events": Array [],
    "instrumentationLibrary": Object {
      "name": "default",
      "version": undefined,
    },
    "kind": 0,
    "links": Array [],
    "name": "External/{name property of client.trackDependency}",
    "parentSpanId": undefined,
    "resource": Resource {
      "attributes": Object {
        "service.name": "unknown_service:/Users/mrickard/.nvm/versions/node/v12.19.0/bin/node",
        "telemetry.sdk.language": "nodejs",
        "telemetry.sdk.name": "opentelemetry",
        "telemetry.sdk.version": "0.23.0",
      },
    },
    "spanContext": undefined,
    "startTime": Array [
      1620165670940,
      0,
    ],
    "status": Object {
      "code": 1,
    },
  },
]
`;

exports[`OpenTelemetryAdapter processes app request and dependency when sent as array of strings 1`] = `Array []`;

exports[`OpenTelemetryAdapter processes app request and dependency when sent as array of strings 2`] = `
Array [
  Object {
    "attributes": Object {
      "name": "node buffer",
      "parentSpanId": "parentId",
    },
    "duration": undefined,
    "endTime": Array [
      1620674372139,
      0,
    ],
    "ended": undefined,
    "events": Array [],
    "instrumentationLibrary": Object {
      "name": "default",
      "version": undefined,
    },
    "kind": 0,
    "links": Array [],
    "name": "node buffer",
    "parentSpanId": undefined,
    "resource": Resource {
      "attributes": Object {
        "service.name": "unknown_service:/Users/mrickard/.nvm/versions/node/v12.19.0/bin/node",
        "telemetry.sdk.language": "nodejs",
        "telemetry.sdk.name": "opentelemetry",
        "telemetry.sdk.version": "0.23.0",
      },
    },
    "spanContext": undefined,
    "startTime": Array [
      1620674371908,
      0,
    ],
    "status": Object {
      "code": 1,
    },
  },
  Object {
    "attributes": Object {
      "name": "select customers proc",
      "parentSpanId": "parentId",
    },
    "duration": undefined,
    "endTime": Array [
      1620674372280,
      0,
    ],
    "ended": undefined,
    "events": Array [],
    "instrumentationLibrary": Object {
      "name": "default",
      "version": undefined,
    },
    "kind": 0,
    "links": Array [],
    "name": "select customers proc",
    "parentSpanId": undefined,
    "resource": Resource {
      "attributes": Object {
        "service.name": "unknown_service:/Users/mrickard/.nvm/versions/node/v12.19.0/bin/node",
        "telemetry.sdk.language": "nodejs",
        "telemetry.sdk.name": "opentelemetry",
        "telemetry.sdk.version": "0.23.0",
      },
    },
    "spanContext": undefined,
    "startTime": Array [
      1620674372049,
      0,
    ],
    "status": Object {
      "code": 1,
    },
  },
  Object {
    "attributes": Object {
      "name": "functionName",
      "parentSpanId": "parentId",
    },
    "duration": undefined,
    "endTime": Array [
      1620674391081,
      545654297,
    ],
    "ended": undefined,
    "events": Array [],
    "instrumentationLibrary": Object {
      "name": "default",
      "version": undefined,
    },
    "kind": 0,
    "links": Array [],
    "name": "functionName",
    "parentSpanId": undefined,
    "resource": Resource {
      "attributes": Object {
        "service.name": "unknown_service:/Users/mrickard/.nvm/versions/node/v12.19.0/bin/node",
        "telemetry.sdk.language": "nodejs",
        "telemetry.sdk.name": "opentelemetry",
        "telemetry.sdk.version": "0.23.0",
      },
    },
    "spanContext": undefined,
    "startTime": Array [
      1620674361060,
      0,
    ],
    "status": Object {
      "code": 1,
    },
  },
]
`;

exports[`OpenTelemetryAdapter processes app request and dependency when sent as array of strings 3`] = `
Array [
  Object {
    "attributes": Object {
      "name": "node buffer",
      "parentSpanId": "parentId",
    },
    "duration": undefined,
    "endTime": Array [
      1620674372139,
      0,
    ],
    "ended": undefined,
    "events": Array [],
    "instrumentationLibrary": Object {
      "name": "default",
      "version": undefined,
    },
    "kind": 0,
    "links": Array [],
    "name": "node buffer",
    "parentSpanId": undefined,
    "resource": Resource {
      "attributes": Object {
        "service.name": "unknown_service:/Users/mrickard/.nvm/versions/node/v12.19.0/bin/node",
        "telemetry.sdk.language": "nodejs",
        "telemetry.sdk.name": "opentelemetry",
        "telemetry.sdk.version": "0.23.0",
      },
    },
    "spanContext": undefined,
    "startTime": Array [
      1620674371908,
      0,
    ],
    "status": Object {
      "code": 1,
    },
  },
  Object {
    "attributes": Object {
      "name": "select customers proc",
      "parentSpanId": "parentId",
    },
    "duration": undefined,
    "endTime": Array [
      1620674372280,
      0,
    ],
    "ended": undefined,
    "events": Array [],
    "instrumentationLibrary": Object {
      "name": "default",
      "version": undefined,
    },
    "kind": 0,
    "links": Array [],
    "name": "select customers proc",
    "parentSpanId": undefined,
    "resource": Resource {
      "attributes": Object {
        "service.name": "unknown_service:/Users/mrickard/.nvm/versions/node/v12.19.0/bin/node",
        "telemetry.sdk.language": "nodejs",
        "telemetry.sdk.name": "opentelemetry",
        "telemetry.sdk.version": "0.23.0",
      },
    },
    "spanContext": undefined,
    "startTime": Array [
      1620674372049,
      0,
    ],
    "status": Object {
      "code": 1,
    },
  },
  Object {
    "attributes": Object {
      "name": "functionName",
      "parentSpanId": "parentId",
    },
    "duration": undefined,
    "endTime": Array [
      1620674391081,
      545654297,
    ],
    "ended": undefined,
    "events": Array [],
    "instrumentationLibrary": Object {
      "name": "default",
      "version": undefined,
    },
    "kind": 0,
    "links": Array [],
    "name": "functionName",
    "parentSpanId": undefined,
    "resource": Resource {
      "attributes": Object {
        "service.name": "unknown_service:/Users/mrickard/.nvm/versions/node/v12.19.0/bin/node",
        "telemetry.sdk.language": "nodejs",
        "telemetry.sdk.name": "opentelemetry",
        "telemetry.sdk.version": "0.23.0",
      },
    },
    "spanContext": undefined,
    "startTime": Array [
      1620674361060,
      0,
    ],
    "status": Object {
      "code": 1,
    },
  },
  Object {
    "attributes": Object {
      "name": "External/{name property of client.trackDependency}",
      "parentSpanId": "12345",
    },
    "duration": undefined,
    "endTime": Array [
      1620165671171,
      0,
    ],
    "ended": undefined,
    "events": Array [],
    "instrumentationLibrary": Object {
      "name": "default",
      "version": undefined,
    },
    "kind": 0,
    "links": Array [],
    "name": "External/{name property of client.trackDependency}",
    "parentSpanId": undefined,
    "resource": Resource {
      "attributes": Object {
        "service.name": "unknown_service:/Users/mrickard/.nvm/versions/node/v12.19.0/bin/node",
        "telemetry.sdk.language": "nodejs",
        "telemetry.sdk.name": "opentelemetry",
        "telemetry.sdk.version": "0.23.0",
      },
    },
    "spanContext": undefined,
    "startTime": Array [
      1620165670940,
      0,
    ],
    "status": Object {
      "code": 1,
    },
  },
]
`;
